app/main.py

```diff
@@ -86,128 +86,114 @@
 }
 
 
-def battle(knightsConfig):
+class Knight:
+    def __init__(self,
+                 name: str,
+                 power: int,
+                 hp: int,
+                 armour: dict,
+                 weapon: dict,
+                 potion: dict) -> None:
+        self.name = name
+        self.hp = hp
+        self.power = power
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = 0
+
+    def preparation(self) -> None:
+        # apply armour
+        for i in self.armour:
+            self.protection += i["protection"]
+
+        # apply weapon
+        self.power += self.weapon["power"]
+
+        # apply potion if exist
+        if self.potion is not None:
+            if "power" in self.potion["effect"]:
+                self.power += self.potion["effect"]["power"]
+
+            if "protection" in self.potion["effect"]:
+                self.protection += self.potion["effect"]["protection"]
+
+            if "hp" in self.potion["effect"]:
+                self.hp += self.potion["effect"]["hp"]
+
+
+def battle(knightsconfig: dict) -> dict:
     # BATTLE PREPARATIONS:
 
     # lancelot
-    lancelot = knightsConfig["lancelot"]
+    lancelot = Knight(knightsconfig["lancelot"]["name"],
```

Comment:
Try to write a separate function for this and use loop to avoid repeating code. You repeat this code four times for different kinghts.